### PR TITLE
DCOS-38126 - list of jobs resolver

### PR DESCRIPTION
**Pull request by Fabs**
This PR is about the jobs resolver who allows us to fetch a list of jobs. 

In order to make it compliant to the Job schema, we needed to do an N + 1 query. We have to fix that soon since it is far from prod ready. https://jira.mesosphere.com/browse/DCOS-38201

Testing this is tricky since it is not connected to any view yet. But you can trust my world on it 😄 or play with the tests to convince yourself.

**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [x] Did you add new unit tests?

### CODE REVIEW

File: plugins/jobs/src/js/data/JobModel.ts [hunk 1]
```diff
@@ -524,17 +527,98 @@ const fieldResolvers = {
...
-      return Observable.of(null);
+    ): Observable<Job[] | null> {
+      const {
+        sortBy = "ID",
+        sortDirection = "ASC",
+        filter = null
+      } = args as JobsQueryArgs;
+
+      const pollingInterval$ = Observable.timer(0, pollingInterval);
+      const responses$ = pollingInterval$.switchMap(fetchJobs);
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: Here we used `exhaustMap` in our first implementation 🤔

File: plugins/jobs/src/js/data/JobModel.ts [hunk 2]
```diff
@@ -524,17 +527,98 @@ const fieldResolvers = {
...
+      return (
+        responses$
+          .map(jobs => filterJobs(jobs, filter))
+          // N+1 request madness
+          .map(jobs => jobs.map(job => fetchJobDetail(job.id)))
+          .switchMap(obs => Observable.combineLatest(obs))
+          .map(requests => requests.map(typeResolvers.Job))
+          .map(jobs =>
+            //TODO Why typescript can't figure out that these value is always
+            // defined?
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: because `sortBy` and `sortDirection` aren't typed, this should work 🤔 
```
+      const {
+        sortBy: SortOption = "ID",
+        sortDirection: SortDirection = "ASC",
+        filter: string | null = null
+      } = args;
```

File: plugins/jobs/src/js/data/JobModel.ts [hunk 3]
```diff
@@ -524,17 +527,98 @@ const fieldResolvers = {
...
+
+      const pollingInterval$ = Observable.timer(0, pollingInterval);
+      const responses$ = pollingInterval$.switchMap(fetchJobs);
+
+      return (
+        responses$
+          .map(jobs => filterJobs(jobs, filter))
+          // N+1 request madness
+          .map(jobs => jobs.map(job => fetchJobDetail(job.id)))
+          .switchMap(obs => Observable.combineLatest(obs))
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: what is this doing? 🤔

File: plugins/jobs/src/js/data/JobModel.ts (hunk 1 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: We used, and then we had that discussion in the Curiosity room with Orlando. switchMap is also not the best solution here. We have to do the pooling ticket Daniel added to the sprint, and that you requested on the http-service.

File: plugins/jobs/src/js/data/JobModel.ts [hunk 4]
```diff
@@ -524,17 +527,98 @@ const fieldResolvers = {
...
+        filter = null
+      } = args as JobsQueryArgs;
+
+      const pollingInterval$ = Observable.timer(0, pollingInterval);
+      const responses$ = pollingInterval$.switchMap(fetchJobs);
+
+      return (
+        responses$
+          .map(jobs => filterJobs(jobs, filter))
+          // N+1 request madness
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/data/JobModel.ts]: I would rather like to see a comment pointing to a JIRA ticket that is about removing the n+1 query

File: plugins/jobs/src/js/data/JobModel.ts (hunk 2 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: You are right. I solved that this way before, and then I rebased and lost that.

Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: We have the `JobsQueryArgs`, they are generated from GraphQL and well typed. You would need to write a [user defined type guard](https://github.com/basarat/typescript-book/blob/master/docs/types/typeGuard.md#user-defined-type-guards) to "cast" it into the right type. I would prefer the generated code over the handwritten TBH, I don't know how you feel about this @juliangieseke

File: plugins/jobs/src/js/data/JobModel.ts [hunk 5]
```diff
@@ -23,8 +25,8 @@ export interface JobsQueryArgs {
 
 export interface JobQueryArgs {
   id: string;
-  filter?: string | null;
-  sortBy?: SortOption | null;
+  filter?: string;
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/data/JobModel.ts]: Why did you remove the null?

File: plugins/jobs/src/js/data/JobModel.ts (hunk 2 above)
Comment (juliangieseke) [reply plugins/jobs/src/js/data/JobModel.ts]: generally yes, but it looks like its already messed up, not sure if we are able to re-generate 😢

File: plugins/jobs/src/js/data/JobModel.ts (hunk 3 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: There is a fiddle here where you can play with the code:
https://jsfiddle.net/fabsn/x5zaf129/23/

The main idea is, you have the first Observable<Jobs[]> (an observable that emits jobs). So you have to map in a request for every job. That is what line 613 makes `Jobs[] => Observable<JobsWithDetail>[]`. But a list of Observables (and not an observable of lists) is not very interesting because the list data structure is boring :). So we use Observable.combineLatest to convert `Observable<JobsWithDetail>[] => Observable<JobsWithDetail[]>`, this is line 614 map part. 

At this step you Have an Observable<Observable<JobsWithDetail[]>>, and you need switchMap (or other strategy) to remove the nesting. This is the switch part of line 614.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 4 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Sure.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 5 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I think that was also an attempt to fix the same problem we discussed above.

But on another pass, I don't think the null should be necessary. I think by saying the parameter value is optional, you pass it or not (I know that is just covering the undefined). I see no point in assigning null to the value if for the filter, or for the sorting, the value of null has no specific meaning.

Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: yeah, I think that's fine with me 🤷‍♂️

File: plugins/jobs/src/js/data/JobModel.ts (hunk 1 above)
Comment (juliangieseke) [reply plugins/jobs/src/js/data/JobModel.ts]: Hm, the discussion was about another delay after an already delayed response finished, not about canceling (or not) an already running request 🤔

Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Yes. But both exhaust and switch have their advantages. Switch will keep dropping and never complete, but exhaust also would have older data for a while until it gets stuck.

We need the polling feature on next sprint.

**Comment by Fabs** [comment]
@DanielMSchmidt @juliangieseke please also do not forget this fix:
+      status: run.status, // TODO: derive from where it comes from before passing to this function

File: plugins/jobs/src/js/data/JobModel.ts (hunk 1 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: Agreed with @Fabs this is not the right place to solve it :)

Comment (juliangieseke) [reply plugins/jobs/src/js/data/JobModel.ts]: My comment wasnt meant to block this PR… 🤷‍♂️

File: plugins/jobs/src/js/data/JobModel.ts (hunk 2 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I actually was not able to type on destruction, I made the item non nullable.

File: plugins/jobs/src/js/data/JobModel.ts [hunk 6]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
+    }
+
+    const direction = sortDirection === "ASC" ? 1 : -1;
+    return result * direction;
+  });
+
+  return jobs;
+}
+
+function sortJobById(a: Job, b: Job): number {
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: 🤚 this _compares_ jobs => `compareJobById`?

File: plugins/jobs/src/js/data/JobModel.ts [hunk 7]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
+  return a.id.localeCompare(b.id);
+}
+
+function sortJobByStatus(a: Job, b: Job): number {
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: 🤚 same as above

File: plugins/jobs/src/js/data/JobModel.ts [hunk 8]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
+  return (
+    JobStates[a.scheduleStatus].sortOrder -
+    JobStates[b.scheduleStatus].sortOrder
+  );
+}
+
+function sortJobByLastRun(a: Job, b: Job): number {
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: 🤚 same as above

File: plugins/jobs/src/js/data/JobModel.ts [hunk 9]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
+      _parent = {},
+      args: GeneralArgs = {},
       _context = {}
-    ): Observable<JobConnection | null> {
-      return Observable.of(null);
+    ): Observable<Job[] | null> {
+      const {
+        sortBy = "ID",
+        sortDirection = "ASC",
+        filter = null
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: 🤚 filter is never `null`, default should be "" (at least from the old implementation)

File: plugins/jobs/src/js/data/JobModel.ts [hunk 10]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
+        responses$
+          .map(jobs => filterJobs(jobs, filter))
+          //TODO We need to remove the N + 1 query
+          // https://jira.mesosphere.com/browse/DCOS-38201
+          .map(jobs => jobs.map(job => fetchJobDetail(job.id)))
+          .switchMap(obs => Observable.combineLatest(obs))
+          .map(requests => requests.map(typeResolvers.Job))
+          .map(jobs =>
+            //TODO Why typescript can't figure out that these value is always
+            // defined?
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: 🎨 if this is fixed, can we remove the comment?

**Comment by juliangieseke** [comment]
@Fabs your issue is because the files have moved, did you rebase or something?

File: plugins/jobs/src/js/data/JobModel.ts (hunk 5 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I changed that because it helps me solve the `blabla || null`

File: plugins/jobs/src/js/data/JobModel.ts (hunk 1 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Are you ok to the switchMap then?

File: plugins/jobs/src/js/data/JobModel.ts [hunk 11]
```diff
@@ -524,17 +527,99 @@ const fieldResolvers = {
...
   Query: {
     jobs(
-      _obj = {},
-      _args: GeneralArgs = {},
+      _parent = {},
+      args: GeneralArgs = {},
       _context = {}
-    ): Observable<JobConnection | null> {
-      return Observable.of(null);
+    ): Observable<Job[] | null> {
```
[the remark this answers is not in the document]
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Hum. I see that, we need to add the countings. I would say ✋. That was an oversight on my side and needs fixing.

File: plugins/jobs/src/js/data/JobModel.ts [hunk 12]
```diff
@@ -17,14 +19,14 @@ export interface GeneralArgs {
 
 export interface JobsQueryArgs {
   filter?: string | null;
-  sortBy?: SortOption | null;
-  sortDirection?: SortDirection | null;
+  sortBy?: SortOption;
```
[the remark this answers is not in the document]
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I agree with your point, but from my idea of how filters and sort work, you could specify no filter and no sorting. I would even like that because sort is expensive. 

In any case, I would not break compatibility, and would them do the types as you propose.

File: plugins/jobs/src/js/data/JobModel.ts [hunk 13]
```diff
@@ -4,6 +4,8 @@ import * as MetronomeClient from "#SRC/js/events/MetronomeClient";
 import JobRunList from "#SRC/js/structs/JobRunList";
 import DateUtil from "#SRC/js/utils/DateUtil";
 import { cleanJobJSON } from "#SRC/js/utils/CleanJSONUtil";
+import JobStates from "#SRC/js/constants/JobStates";
+import JobStatus from "#SRC/js/constants/JobStatus";
```
[the remark this answers is not in the document]
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: ah, ok. That is what broke it them. Next 🤠 dev should fix it :D

File: plugins/jobs/src/js/data/JobModel.ts (hunk 6 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: You are right, this is the compare function and not the sort function. My bad to blind copy and paste.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 7 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Agree

File: plugins/jobs/src/js/data/JobModel.ts (hunk 8 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: yes :D

File: plugins/jobs/src/js/data/JobModel.ts (hunk 9 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: Should, I believe I lost this one on rebase this morning. But you are right.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 10 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I removed that. I rebased it.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (juliangieseke) [reply plugins/jobs/src/js/data/JobModel.ts]: We can guard filtering by checking against empty string

File: plugins/jobs/src/js/types/Job.ts [hunk 14]
```diff
@@ -84,25 +84,25 @@ type Job {
...
-    json: fieldResolvers.json(response),
-    labels: fieldResolvers.labels(response),
-    lastRunStatus: fieldResolvers.lastRunStatus(response),
-    name: fieldResolvers.name(response),
-    jobRuns: fieldResolvers.jobRuns(response),
-    lastRunsSummary: fieldResolvers.lastRunsSummary(response),
-    scheduleStatus: fieldResolvers.scheduleStatus(response),
-    activeRuns: fieldResolvers.activeRuns(response),
-    schedules: fieldResolvers.schedules(response)
+    id: JobFieldResolvers.id(response),
```
Comment (juliangieseke) [inline plugins/jobs/src/js/types/Job.ts]: gnah, this needs to be in the other PR 😩

Comment (juliangieseke) [reply plugins/jobs/src/js/types/Job.ts]: it is 🤔

File: plugins/jobs/src/js/data/JobModel.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: I disagree, it is perfectly fine to have it undefined or null here, why would we want to force the user of the datalayer to always filter or sort?

**Comment by juliangieseke** [comment]
recovered fabs commits here, lets try again on monday…

**Comment by juliangieseke** [comment]
rebased again, this time it should be fine, but there is one failing test we need to investigate… ( I think @DanielMSchmidt can fix them pretty fast - he fixed something similar in the previous PR)

File: plugins/jobs/src/js/data/JobModel.ts (hunk 1 above)
Comment (juliangieseke) [reply plugins/jobs/src/js/data/JobModel.ts]: As discussed offline with orlando: 
```
Rx.Observable
  .concat(Rx.Observable.from([1]))
  .delay(1000)
  .repeat(-1)
```

File: plugins/jobs/src/js/data/JobModel.ts [hunk 15]
```diff
@@ -76,17 +80,119 @@ function isJobQueryArg(arg: any): arg is JobQueryArgs {
...
+        sortDirection = "ASC",
+        filter = null,
+        namespace = null
+      } = args as JobsQueryArgs;
+
+      const pollingInterval$ = Observable.timer(0, pollingInterval);
+      const responses$ = pollingInterval$.exhaustMap(fetchJobs);
+
+      const jobsInNamespace$ = responses$.map(jobs =>
+        filterJobsByNamespace(jobs, namespace)
```
Comment (nLight) [inline plugins/jobs/src/js/data/JobModel.ts]: we could have created a curried function here :trollface: 
```js
const jobsInNamespace$ = responses$.map(getNamespaceJobsFilter(namespace))
```

File: plugins/jobs/src/js/data/JobModel.ts [hunk 16]
```diff
@@ -76,17 +80,119 @@ function isJobQueryArg(arg: any): arg is JobQueryArgs {
...
+      );
+      const count$ = jobsInNamespace$.map(jobs => jobs.length);
+
+      return (
+        jobsInNamespace$
+          .map(jobs => filterJobsById(jobs, filter))
+          //TODO We need to remove the N + 1 query
+          // https://jira.mesosphere.com/browse/DCOS-38201
+          .map(jobs => jobs.map(job => fetchJobDetail(job.id)))
+          .switchMap(obs => Observable.combineLatest(obs))
```
Comment (nLight) [inline plugins/jobs/src/js/data/JobModel.ts]: I'm sorry but when I see `jobs` and `obs` close to each other I can't stop thinking there's a `j` missing. Let's not abbreviate variables 💪 plus if argument is an Observable we may wanna consider adding `$` at the end

File: plugins/jobs/src/js/data/JobModel.ts [hunk 17]
```diff
@@ -76,17 +80,119 @@ function isJobQueryArg(arg: any): arg is JobQueryArgs {
...
+        filterJobsByNamespace(jobs, namespace)
+      );
+      const count$ = jobsInNamespace$.map(jobs => jobs.length);
+
+      return (
+        jobsInNamespace$
+          .map(jobs => filterJobsById(jobs, filter))
+          //TODO We need to remove the N + 1 query
+          // https://jira.mesosphere.com/browse/DCOS-38201
+          .map(jobs => jobs.map(job => fetchJobDetail(job.id)))
```
Comment (nLight) [inline plugins/jobs/src/js/data/JobModel.ts]: Why do we even need N+1 here? Can't we just have a separate stream for the `JobDetailPageContainer`?

File: plugins/jobs/src/js/data/JobModel.ts (hunk 15 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: I would argue that this is unnecessarily complicated to read, but if you like 🤷‍♂️

File: plugins/jobs/src/js/data/JobModel.ts (hunk 17 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: We made two decisions in the past:

1) We want to have one single GraphQL schema
2) We start with a lightweight GraphQL implementation that does not support field resolvers by now

Currently we need to resolve the type `Job` in the index response as part of the `JobsConnection` (Decision 1). This means we need to give the `Job` type in it's entireness to the GraphQL implementation (this would be lazy instead if we had field resolvers) (Decision 2). Because the `Job` type needs more information then the index API endpoint delivers we now need to fetch the detail endpoint to not break the contract of GraphQL.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 15 above)
Comment (nLight) [reply plugins/jobs/src/js/data/JobModel.ts]: no-no, that was more a comment to troll @Fabs, please ignore

Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I actually started with the partially applied function, then I realized the elegance might have a negative impact on comprehension.

File: plugins/jobs/src/js/data/JobModel.ts [hunk 18]
```diff
@@ -176,21 +176,15 @@ export const resolvers = ({
...
-          .switchMap(observable$ => Observable.combineLatest(observable$))
-          .map(requests => requests.map(JobTypeResolver))
-          .map(jobs => sortJobs(jobs, sortBy, sortDirection))
-          .combineLatest(count$, (jobs, totalCount) => ({
-            filteredCount: jobs.length,
-            totalCount,
-            nodes: jobs
-          }))
-      );
+      return jobsInNamespace$
```
Comment (juliangieseke) [inline plugins/jobs/src/js/data/JobModel.ts]: since there is no n+1 madness anymore, we should be able to simply do 
```
responses$
  .map(jobs => JobConnectionTypeResolver(jobs))
  .map(jobConnection => sortJobs(jobConnection, sortBy, sortDirection);
```
which then contains all the needed information, so no mapping over all jobs (here) and `combineLatest` necessary anymore.

File: plugins/jobs/src/js/types/Job.ts [hunk 19]
```diff
@@ -64,26 +66,28 @@ ${JobDockerSchema}
...
   mem: Int!
   name: String!
   schedules: ScheduleConnection!
   scheduleStatus: JobStatus!
 }
 `;
 
-export function JobTypeResolver(response: MetronomeJobDetailResponse): Job {
+export function JobTypeResolver(
+  response: MetronomeJobDetailResponse | MetronomeJobResponse
```
Comment (juliangieseke) [inline plugins/jobs/src/js/types/Job.ts]: 😭

File: plugins/jobs/src/js/types/Job.ts [hunk 20]
```diff
@@ -103,72 +107,102 @@ export function JobTypeResolver(response: MetronomeJobDetailResponse): Job {
...
+          return "SCHEDULED";
+        }
       }
     }
 
     if (
-      scheduleConnection.nodes.length === 0 &&
-      JobRunConnectionTypeResolver(job.activeRuns).nodes.length === 0
+      scheduleConnection.nodes.length === 0 // &&
+      // TODO: clairify this if
```
Comment (juliangieseke) [inline plugins/jobs/src/js/types/Job.ts]: ?

File: plugins/jobs/src/js/types/Job.ts [hunk 21]
```diff
@@ -103,73 +107,99 @@ export function JobTypeResolver(response: MetronomeJobDetailResponse): Job {
...
+  labels(job: MetronomeJobDetailResponse | MetronomeJobResponse): JobLabel[] {
     return Object.entries(job.labels).map(([key, value]) => ({ key, value }));
   },
-  lastRunsSummary(job: MetronomeJobDetailResponse): JobHistorySummary {
-    return JobHistorySummaryTypeResolver(job.history);
+  lastRunsSummary(
+    job: MetronomeJobDetailResponse | MetronomeJobResponse
+  ): JobHistorySummary | null {
+    return isMetronomeJobDetailResponse(job)
+      ? JobHistorySummaryTypeResolver(job.history)
```
Comment (juliangieseke) [inline plugins/jobs/src/js/types/Job.ts]: this needs a small fixup, its `job.history` but `jobs[].historySummary` - content is almost the same tho

Comment (nLight) [reply plugins/jobs/src/js/types/Job.ts]: let's fix this

**Comment by DanielMSchmidt** [comment]
@juliangieseke @nLight This is ready again, could you take another look. Please have in mind that we can change this file, later on, let's not feature creep 🙈

File: src/js/events/MetronomeClient.ts [hunk 22]
```diff
@@ -5,43 +5,15 @@ import { Observable } from "rxjs/Observable";
...
-  id: string;
-  description: string;
   labels: LabelResponse;
   run: {
     cpus: number;
     mem: number;
     disk: number;
     cmd: string;
-    env: {};
+    env: object;
```
Comment (nLight) [inline src/js/events/MetronomeClient.ts]: 👍

Comment (DanielMSchmidt) [reply src/js/events/MetronomeClient.ts]: 👍
